### PR TITLE
fix(core): serialize local thread storage mutations

### PR DESCRIPTION
**Pull request by Kinfe123**
## Summary

- serialize local thread metadata and message-history mutations per storage key
- preserve concurrency between unrelated thread histories
- allow later mutations to continue after a storage operation rejects

## Why

Local thread metadata and message histories are stored as JSON blobs using a read-modify-write cycle. When two mutations overlap, both can read the same old snapshot and then replace the complete blob. The write that finishes last silently removes the other update.

One real timing-dependent case is a slow async storage adapter combined with a fast model response: persistence of the user message is still in progress when persistence of the assistant response begins. Both appends can read the previous history, and after a reload either new message may be missing. Concurrent thread initialization or metadata updates can lose data in the same way.

This change queues the complete read-modify-write transaction by storage key. The first mutation starts immediately, and only another mutation targeting that same key waits. Message writes for different threads still run concurrently, so the fix does not introduce a global storage bottleneck.

## Testing

- added a deterministic regression test that pauses the first metadata write, starts a second initialization, and verifies the second read waits and both threads survive
- added coverage that a rejected mutation does not block the next mutation
- `pnpm --filter @assistant-ui/core test`
- `pnpm lint`
- `pnpm build`

<a href="https://cubic.dev/pr/assistant-ui/assistant-ui/pull/4847?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx [hunk 1]
```diff
@@ -343,64 +368,78 @@ export const createLocalStorageAdapter = (
...
-      }
+      await mutationQueue.run(threadsKey, async () => {
+        const threads = await loadThreadMetadata();
+        const thread = threads.find((t) => t.remoteId === remoteId);
+        if (thread) {
+          thread.status = "regular";
+          await saveThreadMetadata(threads);
+        }
+      });
     },
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx:437]: Addressed in d5c5e36af: added a local updateThreadMetadata helper and reused it across rename, custom metadata, archive state, and generated-title updates.

File: packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx [hunk 2]
```diff
@@ -343,64 +368,78 @@ export const createLocalStorageAdapter = (
...
-      await saveThreadMetadata(filtered);
-      await storage.removeItem(messagesKey(remoteId));
+      await mutationQueue.run(threadsKey, async () => {
+        const threads = await loadThreadMetadata();
+        const filtered = threads.filter((t) => t.remoteId !== remoteId);
+        await saveThreadMetadata(filtered);
+      });
+      await mutationQueue.run(messagesKey(remoteId), () =>
+        storage.removeItem(messagesKey(remoteId)),
+      );
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx]: Addressed in d5c5e36af: compute the message storage key once and reuse it for both queueing and removal.

File: packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx [hunk 3]
```diff
@@ -311,6 +335,7 @@ export const createLocalStorageAdapter = (
 
   const threadsKey = `${prefix}threads`;
   const messagesKey = (threadId: string) => `${prefix}messages:${threadId}`;
+  const mutationQueue = new KeyedMutationQueue();
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/adapters/LocalStorageThreadListAdapter.tsx]: Valid edge case, addressed in 16b87a835. Mutation queues are now shared through a WeakMap keyed by the AsyncStorageLike instance, while the queue keys retain prefix/thread isolation. The regression test now runs the overlapping initializations through two adapters sharing the same storage.

**Review by okisdev — APPROVED** [review]
## review

lgtm as `patch`. verified both directions at runtime: at the head the adapter suite passes 13/13 and the full core suite 492/492, and with `LocalStorageThreadListAdapter.tsx` reverted to main the concurrency regression test fails exactly at the discriminating assertion (the second read-modify-write reads the stale snapshot while the first write is pending, which is the lost-write mechanism).

the queue design holds up under scrutiny: per-key chains keep unrelated threads concurrent, the swallowed tail keeps a rejected mutation from blocking the next one while the caller still sees the rejection, drained keys are deleted from the map so it cannot grow unbounded, and the module-level WeakMap shares one queue per storage object so two adapters over the same storage serialize (the two-adapter test pins this). every read-modify-write site is queued, including delete's messages-key removal, which also closes the delete-versus-late-append resurrection edge.

### shortcomings of the original implementation

- none beyond the reported race; the review iteration already moved the queue from per-adapter to per-storage and extracted the shared metadata helper

### what we modified

- nothing; the PR merges as submitted

### potential further fixes

- two distinct storage wrapper objects over the same underlying store still get separate queues; identity-based sharing is the strongest guarantee available at this layer, worth a doc note only if the pattern shows up in the wild
- reads (`list`/`fetch`/history `load`) stay unqueued and can observe the pre-write blob during a mutation, which matches the previous semantics; queueing them would trade read latency for read-after-write consistency and deserves its own discussion if anyone hits it
